### PR TITLE
Improve video_id extraction in YoutubeLoader

### LINKED ISSUE #4451: YoutubeLoader.from_youtube_url should handle common YT url formats
### Feature request

`YoutubeLoader.from_youtube_url` accepts single URL format. It should be able to handle at least the most common types of Youtube urls out there.

### Motivation

Current video id extraction is pretty naive. It doesn't handle anything than single specific type of youtube url. Any valid but different video address leads to exception.

### Your contribution

I've prepared a PR where i've introduced `.extract_video_id` method. Under the hood it uses regex to find video id in most popular youtube urls. Regex is based on Youtube-dl solution which can be found here: https://github.com/ytdl-org/youtube-dl/blob/211cbfd5d46025a8e4d8f9f3d424aaada4698974/youtube_dl/extractor/youtube.py#L524

### PR DESCRIPTION
**Pull request by KaczuH**
# Improve video_id extraction in `YoutubeLoader`

`YoutubeLoader.from_youtube_url` can only deal with one specific url format. I've introduced `YoutubeLoader.extract_video_id` which can extract video id from common YT urls.

Fixes #4451 

@eyurtsev

### CODE REVIEW

File: langchain/document_loaders/youtube.py [hunk 1]
```diff
@@ -96,6 +97,34 @@ def _load_credentials(self) -> Any:
 
         return creds
 
+YT_URL_RE = re.compile(
```
[the remark this answers is not in the document]
Comment (KaczuH) [reply langchain/document_loaders/youtube.py:100]: > Thank you for getting all the edge cases mapped out and added to the unit tests!

My list is not exhaustive but should be enough for majority of LangChain users. I got these from: https://gist.github.com/rodrigoborgesdeoliveira/987683cfbfcc8d800192da1e73adc486

> Do you think there will be any issues to re-write this to use python

No, it's not a problem. I can do it.

Depends on how wide range of cases do we want to support.

Before final decision please consider:

**This regex is mature solution and handles any YT url LangChain users will probably ever encounter. `urllib.parse` is more readable and maintainable but it'd be difficult to reach the scope regex covers.**

It's stripped down version based on much more elaborate one from Youtube-dl source: https://github.com/ytdl-org/youtube-dl/blob/211cbfd5d46025a8e4d8f9f3d424aaada4698974/youtube_dl/extractor/youtube.py#L524

I've used it as i believe yt-dl has one of the most battle tested extractors available. I doubt it will have to be updated beyond what's in PR.

But i fully agree with readability and maintainability concern. 

What do you think?

[the remark this answers is not in the document]
Comment (eyurtsev) [reply langchain/document_loaders/youtube.py:100]: I agree that the referenced source is battle tested, so it's likely handling all the variations seen in the wild. Do you know if there are accompanying unit tests for the source implementation that we could use in the form of a sequence of URLs? 

Why did you need to strip down the original implementation? Given that the implementation is different from the source, it's hard for me to tell at a glance whether an issue has been introduced.

The other concern with regexps that have such complexity is that they may be prone to regexps attacks.

@vowelparrot   / @dev2049 could you weigh in here with any thoughts?

[the remark this answers is not in the document]
Comment (KaczuH) [reply langchain/document_loaders/youtube.py:100]: > Do you know if there are accompanying unit tests for the source implementation that we could use in the form of a sequence of URLs?

There is only one test i've found but it does not test all cases.
https://github.com/ytdl-org/youtube-dl/blob/211cbfd5d46025a8e4d8f9f3d424aaada4698974/test/test_youtube_misc.py#L14

`YoutubeIE._TESTS` class variable seems to have more urls.
https://github.com/ytdl-org/youtube-dl/blob/211cbfd5d46025a8e4d8f9f3d424aaada4698974/youtube_dl/extractor/youtube.py#L570

I've already gone through them. All cases are in the PR.

> Why did you need to strip down the original implementation?

The original regex is too excessive. It handles any known address redirecting to YT videos (youtuberepeat, various video downloaders, alternative frontends like Invidious etc). I assumed Langchain users don't need that. What i'd expect from `YoutubeLoader.from_youtube_url` (btw this name should be less verbose e.g.`YoutubeLoader.from_url`) to correctly handle most of standard urls.

 > The other concern with regexps that have such complexity is that they may be prone to regexps attacks.
 
I agree.

I'm not gonna push it. I just want to provide as much context as possible to facilitate decision making process.

[the remark this answers is not in the document]
Comment (eyurtsev) [reply langchain/document_loaders/youtube.py:100]: I'll comandeer and implement the non regexp version. I think it means that the code will need to be updated a bit as it won't handle all cases, but I think those updates will be easier to implement

[the remark this answers is not in the document]
Comment (KaczuH) [reply langchain/document_loaders/youtube.py:100]: Sure i get it :) Let me know if i can help.
